pkg/git/cache.go

```diff
@@ -0,0 +1,141 @@
+package git
+
+import (
+	"crypto/sha1"
+	"encoding/hex"
+	"hash"
+	"os"
+	"path"
+	"sync"
+
+	"github.com/git-depend/git-depend/pkg/utils"
+)
+
+// Stores the path of the cache
+type Cache struct {
+	sync.Mutex
+	// root path to the cache
+	root string
+	// repositories maps the URL to the filepath
+	repositories map[string]string
+}
+
+func NewCache(path string) (*Cache, error) {
+	_, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		if err := os.Mkdir(path, 0755); err != nil {
+			return nil, err
+		}
+	}
+	return &Cache{
+		root:         path,
+		repositories: make(map[string]string),
+	}, nil
+}
+
+func (cache *Cache) getNewHasher() hash.Hash {
+	return sha1.New()
+}
+
+// CloneOrUpdate figures out if we need to clone the repo or simply update.
+// Clones are done by taking the shasum of the url and using that as the directory name.
+// Returns the sha and error.
+func (cache *Cache) CloneOrUpdate(url string) (string, error) {
+	h := cache.getNewHasher()
+	h.Write([]byte(url))
+	sha := hex.EncodeToString(h.Sum(nil))
+
+	directory := path.Join(cache.root, sha)
+	_, err := os.Stat(directory)
+	// If directory exists, add it to the map.
+	// If not, create it.
+	if os.IsNotExist(err) {
+		// Create a tmp directory incase something goes wrong.
+		tmp_directory := path.Join(cache.root, "tmp", sha)
+		if err := os.RemoveAll(tmp_directory); err != nil {
+			return "", err
```

Comment:
Not a comment but man do I wish go would let you `return _` for the default value rather than littering static strings everywhere.